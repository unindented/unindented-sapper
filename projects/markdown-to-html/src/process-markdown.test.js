const stringify = require("rehype-stringify");
const processor = require("./process-markdown");

const process = text =>
  new Promise((resolve, reject) => {
    processor()
      .use(stringify)
      .process(text, (err, contents) => {
        if (!err) {
          resolve(contents.contents);
        } else {
          reject(err);
        }
      });
  });

describe("markdown", () => {
  it("processes mermaid blocks", () => {
    const text = `
\`\`\` mermaid
graph LR;
A-->B;
\`\`\`
`;
    return expect(process(text)).resolves.toContain("svg");
  });

  it("converts quotes to smart quotes", () => {
    const text = `
John said, "I saw Lucy at lunch, she told me 'Mary wants you to get some ice cream on your way home'. I think I will get some at Ben and Jerry's, on Gloucester Road."
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("processes abbreviations", () => {
    const text = `
Get the latest news from the [BBC](https://www.bbc.co.uk/) in Stoke and Staffs.

*[BBC]: British Broadcasting Corporation
*[Staffs]: Staffordshire
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("processes youtube iframes", () => {
    const text = `
!(https://www.youtube.com/watch?v=dQw4w9WgXcQ)
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("processes keyboard keys", () => {
    const text = `
To take a screenshot on your Mac, press ||Cmd|| + ||Shift|| + ||3||.
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("processes superscript and subscript text", () => {
    const text = `
Chemical formulas are written using subscripts (e.g. C~6~H~12~O~6~), but atomic isotopes are written using superscripts (e.g. ^13^C, ^131^I, and ^238^U).
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("processes tables", () => {
    const text = `
+---+---+
| A | B |
+---+---+
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("processes math blocks", () => {
    const text = `
$$
\\frac{n!}{k!(n-k)!} = \\binom{n}{k}
$$
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("processes inline math blocks", () => {
    const text = `
In analytical mathematics, _Euler's identity_ is the equality $e^{i\\pi} + 1 = 0$.
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("processes code blocks", () => {
    const text = `
\`\`\`javascript
console.log('Hello, world!')
\`\`\`
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("adds webp alternative to png images", () => {
    const text = `
<img src="logo.png" alt="Logo" />
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });

  it("adds fragment links to headings", () => {
    const text = `
# First-Level Header

## Second-Level Header

### Third-Level Header
`;
    return expect(process(text)).resolves.toMatchSnapshot();
  });
});
