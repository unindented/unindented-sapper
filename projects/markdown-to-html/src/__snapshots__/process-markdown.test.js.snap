// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown adds fragment links to headings 1`] = `
"<h1 id=\\"first-level-header\\">First-Level Header<a class=\\"hash\\" href=\\"#first-level-header\\">#</a></h1>
<h2 id=\\"second-level-header\\">Second-Level Header<a class=\\"hash\\" href=\\"#second-level-header\\">#</a></h2>
<h3 id=\\"third-level-header\\">Third-Level Header<a class=\\"hash\\" href=\\"#third-level-header\\">#</a></h3>"
`;

exports[`markdown adds webp alternative to png images 1`] = `"<picture><source srcset=\\"logo.webp\\" type=\\"image/webp\\"><img src=\\"logo.png\\" alt=\\"Logo\\"></picture>"`;

exports[`markdown converts quotes to smart quotes 1`] = `"<p>John said, “I saw Lucy at lunch, she told me ‘Mary wants you to get some ice cream on your way home’. I think I will get some at Ben and Jerry’s, on Gloucester Road.”</p>"`;

exports[`markdown processes abbreviations 1`] = `"<p>Get the latest news from the <a href=\\"https://www.bbc.co.uk/\\"><abbr title=\\"British Broadcasting Corporation\\">BBC</abbr></a> in Stoke and <abbr title=\\"Staffordshire\\">Staffs</abbr>.</p>"`;

exports[`markdown processes code blocks 1`] = `
"<pre><code class=\\"hljs language-javascript\\"><span class=\\"hljs-built_in\\">console</span>.log(<span class=\\"hljs-string\\">'Hello, world!'</span>)
</code></pre>"
`;

exports[`markdown processes inline math blocks 1`] = `"<p>In analytical mathematics, <em>Euler’s identity</em> is the equality <span class=\\"inlineMath\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math><semantics><mrow><msup><mi>e</mi><mrow><mi>i</mi><mi>π</mi></mrow></msup><mo>+</mo><mn>1</mn><mo>=</mo><mn>0</mn></mrow><annotation encoding=\\"application/x-tex\\">e^{i\\\\pi} + 1 = 0</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.907994em;vertical-align:-0.08333em;\\"></span><span class=\\"mord\\"><span class=\\"mord mathdefault\\">e</span><span class=\\"msupsub\\"><span class=\\"vlist-t\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.824664em;\\"><span style=\\"top:-3.063em;margin-right:0.05em;\\"><span class=\\"pstrut\\" style=\\"height:2.7em;\\"></span><span class=\\"sizing reset-size6 size3 mtight\\"><span class=\\"mord mtight\\"><span class=\\"mord mathdefault mtight\\">i</span><span class=\\"mord mathdefault mtight\\" style=\\"margin-right:0.03588em;\\">π</span></span></span></span></span></span></span></span></span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;\\"></span><span class=\\"mbin\\">+</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;\\"></span><span class=\\"mord\\">1</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;\\"></span><span class=\\"mrel\\">=</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;\\"></span><span class=\\"mord\\">0</span></span></span></span></span>.</p>"`;

exports[`markdown processes keyboard keys 1`] = `"<p>To take a screenshot on your Mac, press <kbd>Cmd</kbd> + <kbd>Shift</kbd> + <kbd>3</kbd>.</p>"`;

exports[`markdown processes math blocks 1`] = `"<div class=\\"math\\"><span class=\\"katex-display\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math><semantics><mrow><mfrac><mrow><mi>n</mi><mo>!</mo></mrow><mrow><mi>k</mi><mo>!</mo><mo>(</mo><mi>n</mi><mo>−</mo><mi>k</mi><mo>)</mo><mo>!</mo></mrow></mfrac><mo>=</mo><mrow><mo fence=\\"true\\">(</mo><mfrac linethickness=\\"0px\\"><mi>n</mi><mi>k</mi></mfrac><mo fence=\\"true\\">)</mo></mrow></mrow><annotation encoding=\\"application/x-tex\\">\\\\frac{n!}{k!(n-k)!} = \\\\binom{n}{k}</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:2.30744em;vertical-align:-0.936em;\\"></span><span class=\\"mord\\"><span class=\\"mopen nulldelimiter\\"></span><span class=\\"mfrac\\"><span class=\\"vlist-t vlist-t2\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:1.37144em;\\"><span style=\\"top:-2.314em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"mord\\"><span class=\\"mord mathdefault\\" style=\\"margin-right:0.03148em;\\">k</span><span class=\\"mclose\\">!</span><span class=\\"mopen\\">(</span><span class=\\"mord mathdefault\\">n</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;\\"></span><span class=\\"mbin\\">−</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;\\"></span><span class=\\"mord mathdefault\\" style=\\"margin-right:0.03148em;\\">k</span><span class=\\"mclose\\">)</span><span class=\\"mclose\\">!</span></span></span><span style=\\"top:-3.23em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"frac-line\\" style=\\"border-bottom-width:0.04em;\\"></span></span><span style=\\"top:-3.677em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"mord\\"><span class=\\"mord mathdefault\\">n</span><span class=\\"mclose\\">!</span></span></span></span><span class=\\"vlist-s\\">​</span></span><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.936em;\\"><span></span></span></span></span></span><span class=\\"mclose nulldelimiter\\"></span></span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;\\"></span><span class=\\"mrel\\">=</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:2.40003em;vertical-align:-0.95003em;\\"></span><span class=\\"mord\\"><span class=\\"mopen delimcenter\\" style=\\"top:0em;\\"><span class=\\"delimsizing size3\\">(</span></span><span class=\\"mfrac\\"><span class=\\"vlist-t vlist-t2\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:1.1075599999999999em;\\"><span style=\\"top:-2.314em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"mord\\"><span class=\\"mord mathdefault\\" style=\\"margin-right:0.03148em;\\">k</span></span></span><span style=\\"top:-3.677em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"mord\\"><span class=\\"mord mathdefault\\">n</span></span></span></span><span class=\\"vlist-s\\">​</span></span><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.686em;\\"><span></span></span></span></span></span><span class=\\"mclose delimcenter\\" style=\\"top:0em;\\"><span class=\\"delimsizing size3\\">)</span></span></span></span></span></span></span></div>"`;

exports[`markdown processes superscript and subscript text 1`] = `"<p>Chemical formulas are written using subscripts (e.g. C<sub>6</sub>H<sub>12</sub>O<sub>6</sub>), but atomic isotopes are written using superscripts (e.g. <sup>13</sup>C, <sup>131</sup>I, and <sup>238</sup>U).</p>"`;

exports[`markdown processes tables 1`] = `"<table><tbody><tr><td colspan=\\"1\\" rowspan=\\"1\\"><p>A</p></td><td colspan=\\"1\\" rowspan=\\"1\\"><p>B</p></td></tr></tbody></table>"`;

exports[`markdown processes youtube iframes 1`] = `"<iframe src=\\"https://www.youtube.com/embed/dQw4w9WgXcQ\\" width=\\"560\\" height=\\"315\\" allowfullscreen frameborder=\\"0\\"></iframe>"`;
